test/runtests.jl

```diff
@@ -389,4 +760,233 @@ using Test
+        @test layout[1][:b] == 4.4
+        @test layout[end][:a] == 3
+        @test layout[end][:b] == 3.3
+        @test layout[:a][1] == 4
+        @test layout[:b][1] == 4.4
+        @test layout[:a][end] == 3
+        @test layout[:b][end] == 3.3
+
+        a_layout = layout.content.contents[:a]
+        b_layout = layout.content.contents[:b]
+
+        AwkwardArray.push!(a_layout, 6)
+        AwkwardArray.push!(b_layout, 6.6)
+        AwkwardArray.end_record!(layout)
+        @test length(layout) == 7
+        @test layout[end][:a] == 6
+        @test layout[end][:b] == 6.6
+        @test layout.index == [3, 4, 0, 0, 1, 2, 5]
+    end
+
+    ### IndexedOptionArray ###################################################
+
+    begin
+        layout = AwkwardArray.IndexedOptionArray(
+            [4, 3, 3, -1, -1, 0],
+            AwkwardArray.PrimitiveArray([1.1, 2.2, 3.3, 4.4, 5.5]),
+        )
+        @test AwkwardArray.is_valid(layout)
+        @test length(layout) == 6
+        @test layout[1] == 5.5
+        @test layout[2] == 4.4
+        @test layout[3] == 4.4
+        @test layout[4] == nothing
+        @test layout[5] == nothing
+        @test layout[6] == 1.1
+        @test layout[2:3] == AwkwardArray.PrimitiveArray([4.4, 4.4])
+        tmp = 0.0
+        for x in layout
+            if x != nothing
+                @test x < 6
+                tmp += x
+            end
+        end
+        @test tmp == 15.4
+
+        AwkwardArray.push!(layout, 6.6)
+        @test length(layout) == 7
+        @test layout[7] == 6.6
+
+        AwkwardArray.push_null!(layout)
+        @test length(layout) == 8
+        @test layout[8] == nothing
+        AwkwardArray.push_null!(layout)
+        @test length(layout) == 9
+        @test layout[9] == nothing
+
+        AwkwardArray.push!(layout, 7.7)
+        @test length(layout) == 10
+        @test layout[10] == 7.7
+        @test layout.index == [4, 3, 3, -1, -1, 0, 5, -1, -1, 6]
+    end
+
+    ### ByteMaskedArray ######################################################
+
+    begin
+        layout = AwkwardArray.ByteMaskedArray(
+            [false, true, true, false, false],
+            AwkwardArray.PrimitiveArray([1.1, 2.2, 3.3, 4.4, 5.5]),
+            valid_when = false,
+        )
+        @test AwkwardArray.is_valid(layout)
+        @test length(layout) == 5
+        @test layout[1] == 1.1
+        @test layout[2] == nothing
+        @test layout[3] == nothing
+        @test layout[4] == 4.4
+        @test layout[5] == 5.5
+        @test layout[4:5] == AwkwardArray.PrimitiveArray([4.4, 5.5])
+        tmp = 0.0
+        for x in layout
+            if x != nothing
+                @test x < 6
+                tmp += x
+            end
+        end
+        @test tmp == 11.0
+
+        AwkwardArray.push!(layout, 6.6)
+        @test length(layout) == 6
+        @test layout[6] == 6.6
+
+        AwkwardArray.push_null!(layout)
+        @test length(layout) == 7
+        @test layout[7] == nothing
```

Comment:
In Python, missing values are represented by `None`, so I implemented them as `nothing::Nothing` here.

However, I see that Julia has a `missing::Missing` singleton, but it is infectious, a bit like `NaN`, but better:

```julia
julia> 5.5 == missing
missing
```

Reply:
cb372e250915ae5da03a6247d7d119eabe5eeb77 replaces `nothing` with `missing`. I don't know: I could go either way. This is supposed to represent option-type data, data that might not be there. Would users expect `nothing` or `missing`? Is there more tooling available for one or the other?

It is a little cumbersome having to avoid testing with `==`, but I can see why it works that way.

Reply:
`missing` has three-way logic, `nothing` is more like `None` in Python.

Basically, `missing` is used more commonly in "[data is missing](https://docs.julialang.org/en/v1/manual/missing/)" context, equivalent to R's `NA` and SQL's `Null`

Reply:
Yeah, so I'm not sure which to use. The `None` values that appear in Awkward Array in Python are intended to mean missing data, and they are a direct translation of Arrow's `null` values. Maybe I could settle this by seeing what Arrow-in-Julia uses.

Reply:
https://github.com/apache/arrow-julia/blob/95efe954784062a115c2a5af15f5c4dbde3fe6f3/src/ArrowTypes/src/ArrowTypes.jl#L45

looks like `Missing` to me!